File: tests/network_operations/ping_test/ping_test.ts
#!/usr/bin/env node
/**
 * Ping test implementation in TypeScript/Node.js.
 * Measures network latency and packet loss to specified targets.
 * Optimized version for better performance using concurrent execution.
 */

import * as fs from 'fs';
import { spawn } from 'child_process';
import * as os from 'os';

interface Config {
    parameters: Parameters;
}

interface Parameters {
    targets: string[];
    packet_count?: number;
    timeout?: number;
}

interface PingResult {
    avg_latency: number;
    min_latency: number;
    max_latency: number;
    packet_loss: number;
    execution_time: number;
    error?: string;
}

interface Summary {
    total_targets: number;
    successful_targets: number;
    failed_targets: number;
    overall_avg_latency: number;
}

interface Results {
    start_time: number;
    targets: { [key: string]: PingResult };
    summary: Summary;
    end_time: number;
    total_execution_time: number;
}

function pingHost(host: string, count: number = 3, timeout: number = 3000): Promise<PingResult> {
    return new Promise((resolve) => {
        const startTime = Date.now();
        
        const isWindows = os.platform() === 'win32';
        const cmd = 'ping';
        const args = isWindows 
            ? ['-n', count.toString(), '-w', timeout.toString(), host]
            : ['-c', count.toString(), '-W', Math.ceil(timeout / 1000).toString(), host];

        const pingProcess = spawn(cmd, args);
        let stdout = '';
        let stderr = '';

        pingProcess.stdout.on('data', (data) => {
            stdout += data.toString();
        });

        pingProcess.stderr.on('data', (data) => {
            stderr += data.toString();
        });

        pingProcess.on('close', (code) => {
            const executionTime = (Date.now() - startTime) / 1000;
            
            if (code !== 0) {
                resolve({
                    avg_latency: Infinity,
                    min_latency: Infinity,
                    max_latency: Infinity,
                    packet_loss: 100.0,
                    execution_time: executionTime,
                    error: stderr || 'Ping failed'
                });
                return;
            }

            const result = parsePingOutput(stdout, isWindows);
            result.execution_time = executionTime;
            resolve(result);
        });

        pingProcess.on('error', (error) => {
            const executionTime = (Date.now() - startTime) / 1000;
            resolve({
                avg_latency: Infinity,
                min_latency: Infinity,
                max_latency: Infinity,
                packet_loss: 100.0,
                execution_time: executionTime,
                error: error.message
            });
        });
    });
}

function parsePingOutput(output: string, isWindows: boolean): PingResult {
    const result: PingResult = {
        avg_latency: 0.0,
        min_latency: 0.0,
        max_latency: 0.0,
        packet_loss: 0.0,
        execution_time: 0.0
    };

    try {
        if (isWindows) {
            // Parse Windows ping output
            const lossMatch = output.match(/(\d+)% loss/);
            if (lossMatch) {
                result.packet_loss = parseFloat(lossMatch[1]);
            }

            // Extract individual ping times
            const timeMatches = output.match(/time[<>=]\s*(\d+)ms/g);
            if (timeMatches) {
                const times = timeMatches.map(match => {
                    const timeMatch = match.match(/(\d+)ms/);
                    return timeMatch ? parseFloat(timeMatch[1]) : 0;
                }).filter(time => time > 0);

                if (times.length > 0) {
                    result.min_latency = Math.min(...times);
                    result.max_latency = Math.max(...times);
                    result.avg_latency = times.reduce((a, b) => a + b, 0) / times.length;
                }
            }

            // Try to get average from summary line
            const avgMatch = output.match(/Average = (\d+)ms/);
            if (avgMatch) {
                result.avg_latency = parseFloat(avgMatch[1]);
            }
        } else {
            // Parse Unix/Linux ping output
            const lossMatch = output.match(/(\d+(?:\.\d+)?)% packet loss/);
            if (lossMatch) {
                result.packet_loss = parseFloat(lossMatch[1]);
            }

            // Parse rtt statistics
            const rttMatch = output.match(/rtt min\/avg\/max\/mdev = ([\d.]+)\/([\d.]+)\/([\d.]+)\/([\d.]+) ms/);
            if (rttMatch) {
                result.min_latency = parseFloat(rttMatch[1]);
                result.avg_latency = parseFloat(rttMatch[2]);
                result.max_latency = parseFloat(rttMatch[3]);
            }
        }
    } catch (error) {
        result.error = 'Failed to parse ping output';
    }

    // If no valid latency was parsed, mark as error
    if (result.avg_latency === 0.0 && result.packet_loss === 100.0) {
        result.error = 'Failed to parse ping output';
    }

    return result;
}

async function runPingBenchmark(params: Parameters): Promise<Results> {
    const startTime = Date.now() / 1000;
    
    const packetCount = params.packet_count || 3; // Reduced for better performance
    const timeout = params.timeout || 3000; // Reduced for better performance
    
    const targets: { [key: string]: PingResult } = {};
    let successfulTargets = 0;
    let failedTargets = 0;
    let totalLatency = 0.0;
    let successfulCount = 0;

    // Execute pings concurrently for better performance
    const pingPromises = params.targets.map(async (target) => {
        console.error(`Pinging ${target}...`);
        const pingResult = await pingHost(target, packetCount, timeout);
        return { target, pingResult };
    });

    // Wait for all pings to complete
    const results = await Promise.all(pingPromises);
    
    // Process results
    for (const { target, pingResult } of results) {
        targets[target] = pingResult;
        
        if (!pingResult.error && pingResult.packet_loss < 100.0) {
            successfulTargets++;
            if (isFinite(pingResult.avg_latency)) {
                totalLatency += pingResult.avg_latency;
                successfulCount++;
            }
        } else {
            failedTargets++;
        }
    }

    const overallAvgLatency = successfulCount > 0 ? totalLatency / successfulCount : 0.0;
    const endTime = Date.now() / 1000;

    return {
        start_time: startTime,
        targets,
        summary: {
            total_targets: params.targets.length,
            successful_targets: successfulTargets,
            failed_targets: failedTargets,
            overall_avg_latency: overallAvgLatency
        },
        end_time: endTime,
        total_execution_time: endTime - startTime
    };
}

async function main() {
    if (process.argv.length < 3) {
        console.error(`Usage: ${process.argv[1]} <config_file>`);
        process.exit(1);
    }

    const configFile = process.argv[2];

    try {
        const configContent = fs.readFileSync(configFile, 'utf8');
        const config: Config = JSON.parse(configContent);

        const results = await runPingBenchmark(config.parameters);
        console.log(JSON.stringify(results, null, 2));
    } catch (error) {
        if (error instanceof Error) {
            console.error(`Error: ${error.message}`);
        } else {
            console.error(`Error: ${error}`);
        }
        process.exit(1);
    }
}

// Always run main for benchmark purposes
main().catch(error => {
    console.error('Unhandled error:', error);
    process.exit(1);
});

File: tests/network_operations/ping_test/ping_test.py
#!/usr/bin/env python3
"""
Ping test implementation in Python.
Measures network latency and packet loss to specified targets.
Optimized version for better performance using concurrent execution.
"""

import json
import sys
import time
import subprocess
import platform
import re
import concurrent.futures
from typing import Dict, List, Optional


def ping_host(host: str, count: int = 3, timeout: int = 3000) -> Dict[str, float]:
    """
    Ping a host and return latency statistics.
    
    Args:
        host: Target host to ping
        count: Number of ping packets to send (reduced for better performance)
        timeout: Timeout in milliseconds (reduced for better performance)
        
    Returns:
        Dictionary with latency statistics
    """
    system = platform.system().lower()
    
    if system == "windows":
        cmd = ["ping", "-n", str(count), "-w", str(timeout), host]
    else:
        cmd = ["ping", "-c", str(count), "-W", str(timeout // 1000), host]
    
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=10)
        
        if result.returncode != 0:
            return {
                "avg_latency": float('inf'),
                "min_latency": float('inf'),
                "max_latency": float('inf'),
                "packet_loss": 100.0,
                "error": result.stderr or "Ping failed"
            }
        
        return parse_ping_output(result.stdout, system)
        
    except subprocess.TimeoutExpired:
        return {
            "avg_latency": float('inf'),
            "min_latency": float('inf'),
            "max_latency": float('inf'),
            "packet_loss": 100.0,
            "error": "Timeout"
        }
    except Exception as e:
        return {
            "avg_latency": float('inf'),
            "min_latency": float('inf'),
            "max_latency": float('inf'),
            "packet_loss": 100.0,
            "error": str(e)
        }


def parse_ping_output(output: str, system: str) -> Dict[str, float]:
    """Parse ping command output and extract statistics."""
    stats = {
        "avg_latency": 0.0,
        "min_latency": 0.0,
        "max_latency": 0.0,
        "packet_loss": 0.0
    }
    
    try:
        if system == "windows":
            # Parse Windows ping output
            # Look for packet loss percentage
            loss_match = re.search(r'(\d+)% loss', output)
            if loss_match:
                stats["packet_loss"] = float(loss_match.group(1))
            
            # Look for latency statistics
            # Find all time values in ms
            time_matches = re.findall(r'time[<>=]\s*(\d+)ms', output)
            if time_matches:
                times = [float(t) for t in time_matches]
                stats["min_latency"] = min(times)
                stats["max_latency"] = max(times)
                stats["avg_latency"] = sum(times) / len(times)
            
            # Alternative: look for Average line
            avg_match = re.search(r'Average = (\d+)ms', output)
            if avg_match:
                stats["avg_latency"] = float(avg_match.group(1))
                
        else:
            # Parse Unix/Linux ping output
            # Look for packet loss
            loss_match = re.search(r'(\d+(?:\.\d+)?)% packet loss', output)
            if loss_match:
                stats["packet_loss"] = float(loss_match.group(1))
            
            # Look for rtt statistics line
            rtt_match = re.search(r'rtt min/avg/max/mdev = ([\d.]+)/([\d.]+)/([\d.]+)/([\d.]+) ms', output)
            if rtt_match:
                stats["min_latency"] = float(rtt_match.group(1))
                stats["avg_latency"] = float(rtt_match.group(2))
                stats["max_latency"] = float(rtt_match.group(3))
    
    except Exception:
        # If parsing fails, set default values
        stats = {
            "avg_latency": float('inf'),
            "min_latency": float('inf'),
            "max_latency": float('inf'),
            "packet_loss": 100.0,
            "error": "Failed to parse ping output"
        }
    
    return stats


def run_ping_benchmark(config: Dict) -> Dict:
    """Run ping benchmark with given configuration using concurrent execution."""
    targets = config.get("targets", ["8.8.8.8"])
    packet_count = config.get("packet_count", 3)  # Reduced for better performance
    timeout = config.get("timeout", 3000)  # Reduced for better performance
    
    results = {
        "start_time": time.time(),
        "targets": {},
        "summary": {
            "total_targets": len(targets),
            "successful_targets": 0,
            "failed_targets": 0,
            "overall_avg_latency": 0.0
        }
    }
    
    total_latency = 0.0
    successful_count = 0
    
    # Execute pings concurrently for better performance
    with concurrent.futures.ThreadPoolExecutor(max_workers=min(10, len(targets))) as executor:
        # Submit all ping tasks
        future_to_target = {
            executor.submit(ping_host, target, packet_count, timeout): target 
            for target in targets
        }
        
        # Collect results as they complete
        for future in concurrent.futures.as_completed(future_to_target):
            target = future_to_target[future]
            print(f"Pinging {target}...", file=sys.stderr)
            
            try:
                ping_result = future.result()
                results["targets"][target] = ping_result
                
                if ping_result.get("error") is None and ping_result["packet_loss"] < 100:
                    results["summary"]["successful_targets"] += 1
                    if ping_result["avg_latency"] != float('inf'):
                        total_latency += ping_result["avg_latency"]
                        successful_count += 1
                else:
                    results["summary"]["failed_targets"] += 1
            except Exception as e:
                results["targets"][target] = {
                    "avg_latency": float('inf'),
                    "min_latency": float('inf'),
                    "max_latency": float('inf'),
                    "packet_loss": 100.0,
                    "error": str(e)
                }
                results["summary"]["failed_targets"] += 1
    
    if successful_count > 0:
        results["summary"]["overall_avg_latency"] = total_latency / successful_count
    
    results["end_time"] = time.time()
    results["total_execution_time"] = results["end_time"] - results["start_time"]
    
    return results


def main():
    """Main entry point for ping test."""
    if len(sys.argv) < 2:
        print("Usage: python ping_test.py <config_file>", file=sys.stderr)
        sys.exit(1)
    
    config_file = sys.argv[1]
    
    try:
        with open(config_file, 'r') as f:
            config = json.load(f)
        
        # Extract parameters from config
        parameters = config.get("parameters", {})
        
        # Run the benchmark
        results = run_ping_benchmark(parameters)
        
        # Output results as JSON
        print(json.dumps(results, indent=2))
        
    except FileNotFoundError:
        print(f"Error: Config file '{config_file}' not found", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in config file: {e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()